### PR TITLE
[Relay][Pass] CanonicalizeCast

### PR DESCRIPTION
**Pull request by vinx13**
In the case that upcast is multi-refed like:
```
data(i32) - cast(i8) -cast(i32) - elemwise_op
                                - elemwise_op
```
if we fuse `cast(i32)` to the previous op, the sub-function after fusion will have 32-bit output, instead of 8-bit output, which has larger memory access overhead.
In such case, it is better to create a copy of `cast(i32)`:
```
data(i32) - cast(i8) - cast(i32) - elemwise_op
                     - cast(i32) - elemwise_op
```
Previously we did special handling in CSE to ignore some common expressions in order to make quantization efficient.
This trick is too fragile for example the recent pr #3233 broke it.
This pass makes it easier as we can transform any expressions into such form that has more efficient fusion result.

cc @tqchen @ZihengJiang @eqy 
also feel free to give a better name for this pass

### CODE REVIEW

**Comment by tqchen** [comment]
@ZihengJiang @eqy please help to review this PR

File: python/tvm/relay/ir_pass.py [hunk 1]
```diff
@@ -652,3 +652,19 @@ def partial_evaluate(expr):
       The output expression.
     """
     return _ir_pass.partial_evaluate(expr)
+
+def canonicalize_expr(expr):
```
Comment (tqchen) [inline python/tvm/relay/ir_pass.py]: as we move toward the new pass API, let us simply remove the expr->expr API here

**Comment by jroesch** [comment]
Can we come up with a better name and put in some docs explaining how this pass works in the text format?

Tianqi wrote this example to me yesterday:

```
def @main(x: int8) {
    %1 = cast(%x, f32)
    %2 = cast(%x, f32)
    %3 = exp(%1)
    %4 = log(%2)
    (%3, 4)
}
```
then
```
def @main(x: int8) {
    %1 = fn (%p1: i8) {
        exp(cast(%p1, f32)
    }
    %3 = %1(%x)
    %2 = fn (%p1: i8) {
        log(cast(%p1, f32)
    }
    %4 = %2(%x)
    (%3, 4)
}
```

**Comment by tqchen** [comment]
@vinx13 please address review comments. @jroesch can you suggest possible alternatives for naming?

File: python/tvm/relay/ir_pass.py (hunk 1 above)
Comment (vinx13) [reply python/tvm/relay/ir_pass.py]: @tqchen @zhiics how should I update unittest to use mod->mod API?

Comment (zhiics) [reply python/tvm/relay/ir_pass.py]: I think you can delete this function here, but just keep the one in transform.py

To use mod->mod API, you can do something like this in the unit test:
```python
mod = _module.Module.from_expr(func)
seq = _transform.Sequential([transform.CanonicalizeExpr(), transform.InferType()])
with _transform.PassContext(opt_level=3):
    mod = seq(mod)
canonicalized_expr = mod[mod.entry_func.name_hint]
```

**Comment by vinx13** [comment]
I have updated this PR. Here are some of the alternative names in my mind: `CanonicalizeCast`, `CanonicalizeFusion`.

**Comment by tqchen** [comment]
CanonicalizeCast seems to be a good name, @jroesch what do you think about it

**Comment by jroesch** [comment]
I think we should use `CannoicalizeCast` or `DuplicateCast` or something since this is a cast focused optimization, its not clear to me whether this is always a performance win or not? is that the case?

**Comment by tqchen** [comment]
I think it will always be a perf win.

**Comment by tqchen** [comment]
Thanks @vinx13 @jroesch this PR is now merged

**Pull request merged**